extension/background.js

```diff
@@ -0,0 +1,290 @@
+    });
+  };
+
+  async report({
+    addonIds,
+    url,
+    redirectUrl,
+    telemetryObject,
+    telemetryValue,
+  }) {
+    // This is the (initial) URL before the redirect.
+    const from = await browser.addonsSearchExperiment.getPublicSuffix(url);
+    // This is the URL after the redirect (or the last one of a chain).
+    const to = await browser.addonsSearchExperiment.getPublicSuffix(
+      redirectUrl
+    );
+
+    if (from === to) {
+      // We do not want to report redirects to same public suffixes. However,
+      // we will report redirects from public suffixes belonging to a same
+      // entity (.e.g., `example.com` -> `example.fr`).
+      //
+      // Known limitation: if a redirect chain starts and ends with the same
+      // public suffix, we won't report any event, even if the chain contains
+      // different public suffixes in between.
+      return;
+    }
+
+    for (const addonId of addonIds) {
+      const addonVersion = await browser.addonsSearchExperiment.getAddonVersion(
+        addonId
+      );
+      const extra = { addonId, addonVersion, from, to };
+
+      this.debug(
+        [
+          `recording event: method=${TELEMETRY_METHOD_ETLD_CHANGE}`,
+          `object=${telemetryObject} value=${telemetryValue}`,
+          `extra=${JSON.stringify(extra)}`,
+        ].join(" ")
+      );
+      browser.telemetry.recordEvent(
+        TELEMETRY_CATEGORY,
+        TELEMETRY_METHOD_ETLD_CHANGE,
+        telemetryObject,
+        telemetryValue,
+        extra
+      );
+    }
+  }
+
+  // This method is used when we stop following a request (ID). We report
+  // information if needed, then the requestId is removed from the map of
+  // request IDs to follow. In addition, when there is no request IDs to
+  // follow, we also remove the (wildcard) follow listener.
+  async unfollowRequest({ requestId }) {
+    if (this.requestIdsToFollow.has(requestId)) {
+      const { addonIds, chain } = this.requestIdsToFollow.get(requestId);
+
+      await this.report({
+        addonIds,
+        url: chain[0],
+        redirectUrl: chain[chain.length - 1],
+        telemetryObject: TELEMETRY_OBJECT_OTHER,
+        telemetryValue: TELEMETRY_VALUE_SERVER,
+      });
+
+      this.debug(`stop following requestId=${requestId}`);
+      this.requestIdsToFollow.delete(requestId);
+    }
+
+    if (this.requestIdsToFollow.size === 0) {
+      this.debug(`removing follow listener`);
+      browser.webRequest.onBeforeRequest.removeListener(this.followHandler);
+    }
+  }
+
+  getAddonIdsForUrl(url) {
+    for (const pattern of Object.keys(this.matchPatterns)) {
+      const [urlPrefix] = pattern.split("*");
+
+      if (url.startsWith(urlPrefix)) {
+        return this.matchPatterns[pattern];
+      }
+    }
+
+    return [];
+  }
+
+  debug(message) {
+    if (this.debugMode) {
+      console.debug(message);
+    }
+  }
+}
+
+browser.runtime.onInstalled.addListener(({ temporary: debugMode }) => {
```

Comment:
The `onInstalled` event doesn't fire once the add-on has been installed. Consequently the add-on wouldn't work any more. The `AddonSearchExperiment` construction should be unconditional.

Reply:
ha, shoot. Well in this case, I am going to remove that and I'll toggle the boolean manually...